config/application.rb

```diff
@@ -37,6 +37,14 @@ class Application < Rails::Application
     config.time_zone = "Central Time (US & Canada)"
     config.eager_load_paths << Rails.root.join("lib")
 
+    if Setting.validate_env_vars == "1"
+      Setting.require_keys!(:app_url_protocol, :app_url_host, :app_url_port)
+    end
```

Comment:
I don't love this extra flag (validate_env_vars) which changes whether the keys are required or not. What would it take for us to always make the keys optional? For example, maybe we do this:

1. If any of these vars are set then we require that all 3 are set. But you can leave all 3 blank and the app still works.
2. Would this mean that then within `open_ai.rb` we just add a conditional so that it falls back to file_data_url() if the url_* details are not provided.
3. It would be nice to have a conditional only in a single place so can we avoid also adding a conditional inside `_message.html`? Maybe you make `document_image_url` be smart and it simply return the path if the url_* details are unknown?

Reply:
yeah it's a bit convoluted as is. I just did that optional piece to avoid it validating during the step to precompile assets. I mainly wanted to have this validation to make the app "fail fast" if the necessary values to create the url weren't present. If we fall back to the path or the base64 path then it should be ok since it still works with base64 urls. I'll update this branch